### ISSUE TITLE
Contextual fields instead of redirect pages

### ISSUE THREAD

**Issue by leofeyer** [contao/contao#204]
As discussed in contao/core-bundle#1592, we want to use contextual fields instead of redirect pages in some modules. This would affect the following core modules:

**Personal data**
  - explanation text before submitting the form
  - confirmation text after submitting the form

**Registration**
  - explanation text before submitting the form
  - confirmation text after submitting the form
  - error text in case the opt-in link is invalid (or just the default error message?)
  - confirmation text if the opt-in was successful

**Change password**
  - explanation text before submitting the form
  - confirmation text after submitting the form

**Lost password**
  - explanation text before submitting the "request password" form
  - confirmation text after submitting the "request password" form
  - error text in case the opt-in link is invalid (or just the default error message?)
  - explanation text before submitting the "change password" form
  - confirmation text after submitting the "change password" form

**Newsletter subscribe**
  - explanation text before submitting the "subscribe" form
  - confirmation text after submitting the "subscribe" form
  - error text in case the opt-in link is invalid (or just the default error message?)
  - confirmation text if the opt-in was successful

**Newsletter unsubscribe**
  - explanation text before submitting the "unsubscribe" form
  - confirmation text after submitting the "unsubscribe" form

@contao/developers Do you agree? Did I forget anything?

**Comment by asaage** [comment]
> Did I forget anything?

maybe **comments subscriptions**

**Comment by fritzmg** [comment]
Generally I do not like this approach, because any textual content should usually be editable by editors - and modules typically aren't.

However, I do not have a better idea right now 😉

**Comment by asaage** [comment]
> any textual content should usually be editable by editors

It's not about the editing of text is it?
I think the messages would still originate from the language files.

**Comment by Toflar** [comment]
Looks okay to me, I just think we should allow for simple tokens in those fields. So you could go something like `Thank you ##firstname## for your subscription`.

**Comment by fritzmg** [comment]
> It's not about the editing of text is it?
> I think the messages would still originate from the language files.

I don't think so? Or do I misunderstand it? Isn't this about adding text fields to the modules which will be displayed in the front end at various stages of the process for those modules?

**Comment by asaage** [comment]
You are right - i messed it up.
Well that makes using these modules in different languages quite cubersome i guess :anguished:

**Comment by leofeyer** [comment]
It cannot get any more cumbersome than it is right now with having to create three additional pages with articles and content elements for one "request password" module!

> any textual content should usually be editable by editors - and modules typically aren't

We have never recommended such a practice! Modules contain textual content as well (headlines, e-mail texts, images, etc.), so if you want your editors to edit this content, just enable the necessary fields in the group settings.

Also, I do not think that module contents such as e-mail texts or confirmation texts are changed very often. They are set up initially and then remain the same in most cases.

**Comment by fritzmg** [comment]
Well it's more a personal preference anyway ;)

**Comment by dmolineus** [comment]
> As discussed in contao/core-bundle#1592, we want to use contextual fields instead of redirect pages in some modules.

This wouldn't replace the redirect page but provide an alternative as I remember the last Contao Call right, wouldn't it?

> Did I forget anything?

A regular form, created in the form generator should be adjusted as well.

**Comment by leofeyer** [comment]
> This wouldn't replace the redirect page but provide an alternative as I remember the last Contao Call right, wouldn't it?

Contextual fields and redirect pages will coexist in Contao 4, because removing them would be a BC break. However, they will be deprecated and they will be removed in Contao 5.

**Comment by fritzmg** [comment]
That would limit the amount of layouting you can do on a success page though.

**Comment by dmolineus** [comment]
> Contextual fields and redirect pages will coexist in Contao 4, because removing them would be a BC break. However, they will be deprecated and they will be removed in Contao 5.

So I got it wrong. Otherwise I would have voted against it as I already have such requirements where a jump to page is useful:

- Module is placed in the footer/somewhere in the layout where the confirmation message wouldn't be noticed very well
- Required content of redirect page wouldn't fit in the layout of the actual module
- Target tracking based on tracking the redirect page access (Yeh, I know they could be visited manually)

I definitely see the use case for such message but I hope it will be added as an alternative instead of removing the redirect pages if Contao 5 will be released eventually.

**Comment by Toflar** [comment]
Yeah imho deprecating makes no sense. They can happily co-exist because they cover different use-cases.

**Comment by leofeyer** [comment]
> Module is placed in the footer/somewhere in the layout where the confirmation message wouldn't be noticed very well

@dmolineus What kind of module?

**Comment by frontendschlampe** [comment]
Actually we're updating our extension https://github.com/hofff/contao-formtools and in version 3.0.0-rc1 we've added the feature for an alternative message within a normal form if no redirect page is selected. In the form settings you can place an individual text for success and error message.

We will make some testings now and we hope, that we can release a stable version soon.

**Comment by leofeyer** [comment]
I do not support having three redirect pages **and** three contextual fields in the "lost password" module, just so people can choose between two concepts. That is overkill IMHO.

**Comment by leofeyer** [comment]
As discussed in Mumble on December 20th, we want to have contextual fields for all module states and additional jumpTo pages for selected states. For instance in the "lost password" module:

- State 1 (please enter your e-mail address to request a new password): contextual field
- State 2 (the opt-in link has been clicked -> please enter at least 8 characters): contextual field
- State 3 (your password has been changed): contextual field and optional jumpTo page to redirect the user to the login page

Side note: contextual fields for forms were never an issue.

**Comment by aschempp** [comment]
what's a contextual field to you?

**Comment by leofeyer** [comment]
A textarea with TinyMCE.

**Comment by JLBA** [comment]
> A textarea with TinyMCE.
Finally : some {insert-tags} we can modify the content ?
